aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/ApplicationLifeCycleHelper.java

```diff
@@ -0,0 +1,161 @@
+    }
+
+    @Override
+    public void onActivityResumed(final Activity activity) {
+        Log.d(LOG_TAG, "onActivityResumed " + activity.getLocalClassName());
+        activityLifecycleStateMap.put(activity, "resumed");
+    }
+
+    @Override
+    public void onActivityPaused(final Activity activity) {
+        Log.d(LOG_TAG, "onActivityPaused " + activity.getLocalClassName());
+        activityLifecycleStateMap.put(activity, "paused");
+    }
+
+    @Override
+    public void onActivityStopped(final Activity activity) {
+        Log.d(LOG_TAG, "onActivityStopped " + activity.getLocalClassName());
+        // When the activity is stopped, we remove it from the lifecycle state map since we
+        // no longer consider it keeping a session alive.
+        activityLifecycleStateMap.remove(activity);
+    }
+
+    @Override
+    public void onActivitySaveInstanceState(final Activity activity, final Bundle outState) {
+        Log.d(LOG_TAG, "onActivitySaveInstanceState " + activity.getLocalClassName());
+    }
+
+    @Override
+    public void onActivityDestroyed(final Activity activity) {
+        Log.d(LOG_TAG, "onActivityDestroyed " + activity.getLocalClassName());
+        // Activity should not be in the activityLifecycleStateMap any longer.
+        if (activityLifecycleStateMap.containsKey(activity)) {
+            Log.wtf(LOG_TAG, "Destroyed activity present in activityLifecycleMap!?");
+            activityLifecycleStateMap.remove(activity);
+        }
+    }
+
+    /**
+     * Call this method when your Application trims memory.
+     * @param level the level passed through from Application.onTrimMemory().
+     */
+    public void handleOnTrimMemory(final int level) {
+        Log.d(LOG_TAG, "onTrimMemory " + level);
+        // If no activities are running and the app has gone into the background.
+        if (level >= Application.TRIM_MEMORY_UI_HIDDEN) {
+            checkForApplicationEnteredBackground();
+        }
+    }
+
+    class ScreenOffReceiver extends BroadcastReceiver {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            checkForApplicationEnteredBackground();
+        }
+    }
+
+    /**
+     * Called back when your application enters the Foreground.
+     */
+    protected abstract void applicationEnteredForeground();
+
+    /**
+     * Called back when your application enters the Background.
+     */
+    protected abstract void applicationEnteredBackground();
+
+    /**
+     * Called from onActivityCreated and onActivityStarted to handle when the application enters
+     * the foreground.
+     */
+    private void handleOnCreateOrOnStartToHandleApplicationEnteredForeground() {
+        // if nothing is in the activity lifecycle map indicating that we are likely in the background, and the flag
+        // indicates we are indeed in the background.
+        if (activityLifecycleStateMap.size() == 0 && !inForeground) {
```

Comment:
When this class first runs, you won't know the right value of `inForeground` for sure. Did this code start executing when the initial state was in foreground? Or in background?

So, I propose that you change this from a `boolean` to an `enum`, that looks like:
```
enum ForegroundState {
    FOREGROUND,
    BACKGROUND,
    UNKNOWN;
}
```